### PR TITLE
Remove STOP_WAITING frame

### LINKED ISSUE #66: Remove STOP_WAITING
@ianswett suggests that maybe this can be removed.  He didn't say why.

### PR DESCRIPTION
**Pull request by ianswett**
First pass at removing the STOP_WAITING frame.  Intends to resolve #66 .

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1209,13 +1208,13 @@ well as which packets are considered missing.  The ACK frame contains between 1
...
-raising the Least Unacked packet number at the receiver.  A sender of an ACK
-frame thus reports only those ACK blocks between the received Least Unacked and
-the reported Largest Acknowledged packet numbers.  The endpoint SHOULD raise the
-Least Unacked communicated via future STOP_WAITING frames to the most recently
-received Largest Acknowledged.
+sender, the sender SHOULD track the ack ranges in acks that have been
+acked and no longer include them in future ACKs.  In order to handle cases
+where the receiver is not sending data, and hence does not expect it's acks to
+be acked, it may bundle a PING frame with an ack approximately once per RTT to
+gain confirmation of it's ACKs being delivered.  The receiver SHOULD track no
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: Done.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -1209,13 +1208,13 @@ well as which packets are considered missing.  The ACK frame contains between 1
...
+more than 255 ack ranges because they are not expressible on the wire, and 
+MAY limit the number of ack ranges to a smaller value.
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: Ok, changed the wording substantially.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -1209,13 +1208,13 @@ well as which packets are considered missing.  The ACK frame contains between 1
...
 
 To limit the ACK blocks to the ones that haven't yet been received by the
-sender, the sender periodically sends STOP_WAITING frames that signal the
-receiver to stop acknowledging packets below a specified sequence number,
-raising the Least Unacked packet number at the receiver.  A sender of an ACK
-frame thus reports only those ACK blocks between the received Least Unacked and
-the reported Largest Acknowledged packet numbers.  The endpoint SHOULD raise the
-Least Unacked communicated via future STOP_WAITING frames to the most recently
-received Largest Acknowledged.
+sender, the sender SHOULD track the ack ranges in acks that have been
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: Thanks.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -1208,14 +1207,16 @@ Receivers send ACK frames to inform senders which packets they have received, as
...
 
-To limit the ACK blocks to the ones that haven't yet been received by the
-sender, the sender periodically sends STOP_WAITING frames that signal the
-receiver to stop acknowledging packets below a specified sequence number,
-raising the Least Unacked packet number at the receiver.  A sender of an ACK
-frame thus reports only those ACK blocks between the received Least Unacked and
-the reported Largest Acknowledged packet numbers.  The endpoint SHOULD raise the
-Least Unacked communicated via future STOP_WAITING frames to the most recently
-received Largest Acknowledged.
+To limit ACK blocks to those that haven't yet been received by the sender,
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: haven't --> have not

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -1208,14 +1207,16 @@ Receivers send ACK frames to inform senders which packets they have received, as
...
+the receiver SHOULD track which ACK frames have been acknowledged by its peer.
+Once an ACK frame has been acknowledged, the packets it acknowledges SHOULD not
+be acknowledged again.  In order to handle cases where the receiver is not
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: "In order to" --> "To"

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -1208,14 +1207,16 @@ Receivers send ACK frames to inform senders which packets they have received, as
...
+sending retransmittable data, and hence does not expect to receive acks, it
+MAY bundle a PING frame with an ack at most once per RTT to gain confirmation
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: "with an ack" --> "with an ACK frame"

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -1208,14 +1207,16 @@ Receivers send ACK frames to inform senders which packets they have received, as
...
+of ACKs being delivered.  In order to limit receiver state or the size of ACK
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: "In order to" --> "To"

Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: "of ACKs being delivered" --> "of ACK frames being delivered"

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -1208,14 +1207,16 @@ Receivers send ACK frames to inform senders which packets they have received, as
...
+frames, a receiver MAY limit the number of ack ranges and stop acknowledging
+packets without getting an ACK acked, with the knowledge this may slightly
+increase spurious retransmits.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I would suggest using "ACK frames" for the noun and "acknowledge" for the verb in general. Suggestion on this particular sentence: see if you can break it up into two sentences. It's a bit of a run-on.

**Comment by ianswett** [comment]
Also resolves #178

**Comment by martinthomson** [comment]
I have updated this based on Jana's comments, and rebased.  I'll merge this as discussed.  @ianswett, feel free to post comments here and I'll fix them directly.

**Comment by ianswett** [comment]
Thanks Martin, looks great!

**Pull request merged**